### PR TITLE
#64: Fix __hash__ implementation for schema classes

### LINKED ISSUE #64: Fix hash and eq function in schema
# Background

Our current Schema classes have some issues with their implementation of `__hash__` and are, in fact, only immutable structs, this is perfect for dataclasses.

The DBObjectName classes should be fine and should be normal classes.

### PR DESCRIPTION
**Pull request by tkilias**
Fixes #64

### CODE REVIEW

File: doc/changes/changes_0.1.0.md [hunk 1]
```diff
@@ -10,4 +10,5 @@
 ### Bug Fixes
 
   - #43: Corrected the git hooks path
+  - #64: Fix `__hash__ implementation for schema classes and add some validations
```
Comment (tomuben) [inline doc/changes/changes_0.1.0.md:13]:
```suggestion
  - #64: Fix `__hash__` implementation for schema classes and add some validations
```

File: exasol_data_science_utils_python/schema/column_name.py [hunk 2]
```diff
@@ -30,3 +30,6 @@ def __eq__(self, other):
 
     def __repr__(self):
         return generate_repr_for_object(self)
+
+    def __hash__(self):
+        return hash((self.name, self._table_name))
```
Comment (tomuben) [inline exasol_data_science_utils_python/schema/column_name.py:35]: I have bad experience with having manual hashing functions, when you add members you easily forget to change the hash function and then error searching can cosume much time.
```suggestion
        return hash(self.__dict__.values(self.name, self._table_name))
```

File: exasol_data_science_utils_python/schema/dbobject.py [hunk 3]
```diff
@@ -25,3 +25,6 @@ def __eq__(self, other):
 
     def __repr__(self):
         return generate_repr_for_object(self)
+
+    def __hash__(self):
+        return hash(self._name)
```
Comment (tomuben) [inline exasol_data_science_utils_python/schema/dbobject.py:30]: same as above, maybe add a safeguard check here to not forget to adjust if needed:
```suggestion
        assert(len(self.__dict__.keys()) == 1) #Update here if necessary
        return hash(self._name)
```

File: exasol_data_science_utils_python/schema/table_like.py [hunk 4]
```diff
@@ -17,6 +17,11 @@ class TableLike(DBObject[NameType], ABC):
     def __init__(self, name: NameType, columns: List[Column]):
         super().__init__(name)
         self._columns = columns
+        if len(self._columns)==0:
+            raise ValueError("At least on column needed.")
```
Comment (tomuben) [inline exasol_data_science_utils_python/schema/table_like.py:21]:
```suggestion
            raise ValueError("At least one column needed.")
```

File: tests/unit_tests/schema/test_column.py [hunk 5]
```diff
@@ -20,3 +20,37 @@ def test_set_new_name_fail():
...
+
+def test_hash_inequality_name():
+    column1 = Column(ColumnNameBuilder.create("abc"), ColumnType("INTEGER"))
+    column2 = Column(ColumnNameBuilder.create("def"), ColumnType("INTEGER"))
+    assert hash(column1) != hash(column2)
+
+
+def test_hash_inequality_type():
+    column1 = Column(ColumnNameBuilder.create("abc"), ColumnType("INTEGER"))
+    column2 = Column(ColumnNameBuilder.create("def"), ColumnType("VARCHAR"))
```
Comment (tomuben) [inline tests/unit_tests/schema/test_column.py:54]: shouldn't be:
```suggestion
    column2 = Column(ColumnNameBuilder.create("abc"), ColumnType("VARCHAR"))
```
?

File: tests/unit_tests/schema/test_table.py [hunk 6]
```diff
@@ -29,3 +49,47 @@ def test_columns_list_is_immutable():
...
+
+def test_inequality_name():
+    table1 = Table(TableNameImpl("table1"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    table2 = Table(TableNameImpl("table2"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    assert table1 != table2
+
+
+def test_inequality_columns():
+    table1 = Table(TableNameImpl("table1"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    table2 = Table(TableNameImpl("table2"),
```
Comment (tomuben) [inline tests/unit_tests/schema/test_table.py:68]: shouldn't be
```suggestion
    table2 = Table(TableNameImpl("table1"),
```
?

File: tests/unit_tests/schema/test_table.py [hunk 7]
```diff
@@ -29,3 +49,47 @@ def test_columns_list_is_immutable():
...
+
+def test_hash_inequality_name():
+    table1 = Table(TableNameImpl("table1"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    table2 = Table(TableNameImpl("table2"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    assert hash(table1) != hash(table2)
+
+
+def test_hash_inequality_columns():
+    table1 = Table(TableNameImpl("table1"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    table2 = Table(TableNameImpl("table2"),
```
Comment (tomuben) [inline tests/unit_tests/schema/test_table.py:90]: same as above

File: tests/unit_tests/schema/test_view.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+def test_inequality_name():
+    view1 = View(ViewNameImpl("view1"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    view2 = View(ViewNameImpl("view2"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    assert view1 != view2
+
+
+def test_inequality_columns():
+    view1 = View(ViewNameImpl("view1"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    view2 = View(ViewNameImpl("view2"),
```
Comment (tomuben) [inline tests/unit_tests/schema/test_view.py:68]: shouldn't be:
```suggestion
    view2 = View(ViewNameImpl("view1"),
```
?

File: tests/unit_tests/schema/test_view.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+def test_hash_inequality_name():
+    view1 = View(ViewNameImpl("view1"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    view2 = View(ViewNameImpl("view2"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    assert hash(view1) != hash(view2)
+
+
+def test_hash_inequality_columns():
+    view1 = View(ViewNameImpl("view1"), [Column(ColumnNameBuilder.create("column"), ColumnType("INTEGER"))])
+    view2 = View(ViewNameImpl("view2"),
```
Comment (tomuben) [inline tests/unit_tests/schema/test_view.py:90]: same as above

**Review by tomuben — CHANGES_REQUESTED** [review]
check my comments

File: exasol_data_science_utils_python/schema/dbobject.py (hunk 3 above)
Comment (tkilias) [reply exasol_data_science_utils_python/schema/dbobject.py]: replaced with generate_hash_for_object

File: tests/unit_tests/schema/test_view.py (hunk 9 above)
Comment (tkilias) [reply tests/unit_tests/schema/test_view.py]: done

File: tests/unit_tests/schema/test_table.py (hunk 7 above)
Comment (tkilias) [reply tests/unit_tests/schema/test_table.py]: done

File: exasol_data_science_utils_python/schema/column_type.py [hunk 10]
```diff
@@ -1,74 +1,42 @@
...
+                             argname="characterSet")
+        typeguard.check_type(value=self.withLocalTimeZone,
+                             expected_type=Optional[bool],
+                             argname="withLocalTimeZone")
+        typeguard.check_type(value=self.fraction,
+                             expected_type=Optional[int],
+                             argname="fraction")
+        typeguard.check_type(value=self.srid,
+                             expected_type=Optional[int],
+                             argname="srid")
```
[the remark this answers is not in the document]
Comment (tkilias) [reply exasol_data_science_utils_python/schema/column_type.py]: done

File: tests/unit_tests/utils/test_hash_generation_for_object.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    test1 = ObjectWithNotHashable(test=NotHashable(1))
+    test2 = ObjectWithNotHashable(test=NotHashable(2))
+
+    result1 = generate_hash_for_object(test1)
+    result2 = generate_hash_for_object(test2)
+    assert result1 == result2
+
+
+@dataclasses.dataclass
+class ObjectWithCycle:
```
Comment (tomuben) [inline tests/unit_tests/utils/test_hash_generation_for_object.py:138]: Actually, I think ObjectWithDuplicates would be a better name

Comment (tkilias) [reply tests/unit_tests/utils/test_hash_generation_for_object.py:138]: I choese this name, because the test this class corresponds to , is about cycles

Comment (tomuben) [reply tests/unit_tests/utils/test_hash_generation_for_object.py:138]: Ok that's fine

**Review by tomuben — APPROVED** [review]

File: exasol_data_science_utils_python/utils/hash_generation_for_object.py [hunk 12]
```diff
@@ -1,22 +1,27 @@
...
-def generate_hash_for_object(obj):
+def generate_hash_for_object(obj: Any) -> int:
     return hash(tuple(_hash_object(v, set())
                       for k, v in sorted(obj.__dict__.items())))
 
 
 def _hash_object(obj: Any, already_seen: Set[int]) -> int:
+    """
+    This function generates a hash value for the object using either the objects __hash__ method,
+    or by inspecting it, if it is a list or list. For non-hashable objects it returns 0.
```
Comment (tomuben) [inline exasol_data_science_utils_python/utils/hash_generation_for_object.py:12]:
```suggestion
    or by inspecting it, if it is a list or dict. For non-hashable objects it returns 0.
```

**Pull request merged**